---
Title: "Bioinformatics for Metagenomics in the Diagnostic Laboratory- Linux & Flowcraft" \
Author: "C I Mendes" \
Date: "October, 2018" 
---

## Overview

This practical session will introduce state-of-the art analysis of Illumina paired-end shotgun metagenomic data in the
context of the Diagnostic Laboratory. Participants will learn the basis of interaction with Linux Operating System 
with Shell commands through a terminal windows, and how to use the [Flowcraft](https://github.com/assemblerflow/flowcraft)
tool to build and execute pipelines to asses microbial composition, resistance gene detection and typing.

## Target audience

Participants of ESCMID's capacity-building workshop for Metagenomics in the Diagnostic Laboratory

## Learning objectives

  * Basic Concepts of Linux and Shell Commands
  
  * Basic Concepts of Nextflow and Software Containerization

  * Overview of FlowCraft 

  * Building a Pipeline with FlowCraft - Recipes and Build Mode

  * Running a Pipeline - Inspector Mode
  
  * Obtaining Interactive Reports 


## Plannification

### 1. [Linux Fundamentals](Linux_Fundamentals.md)

####    1.1 What is the Shell and Why do you need it?

####    1.2 First Commands

####    1.3 What Else is Out There

####    1.4 Useful Material

### 2. [Nextflow and Software Containers](Nextflow_containers.md)

####    2.1 Reproducibility of a Program

####    2.2 Software Containers

####    2.3 The Nextflow 

### 3. [FlowCraft](Flowcraft.md)

####    3.1 Overview

####    3.2 Installation and Dependencies

####    3.3 Building a pipeline

#####       3.3.1 The build mode

#####       3.3.2 Recipes

####    3.4 Running a Pipeline

#####       3.4.1 The Inspector mode

####    3.5 Obtaining a Report